src/iocp/afd.rs

```diff
@@ -0,0 +1,608 @@
+        afd_events: AfdPollMask,
+    ) -> io::Result<()> {
+        const IOCTL_AFD_POLL: u32 = 0x00012024;
+
+        // Lock the packet.
+        if !packet.get().try_lock() {
+            return Err(io::Error::new(
+                io::ErrorKind::WouldBlock,
+                "packet is already in use",
+            ));
+        }
+
+        // Set up the AFD poll info.
+        let poll_info = unsafe {
+            let poll_info = Pin::into_inner_unchecked(packet.get().afd_info()).get();
+
+            // Initialize the AFD poll info.
+            (*poll_info).exclusive = false.into();
+            (*poll_info).handle_count = 1;
+            (*poll_info).timeout = std::i64::MAX;
+            (*poll_info).handles[0].handle = base_socket as HANDLE;
+            (*poll_info).handles[0].status = 0;
+            (*poll_info).handles[0].events = afd_events;
+
+            poll_info
+        };
+
+        let iosb = T::into_ptr(packet).cast::<IO_STATUS_BLOCK>();
+        // Set Status to pending
+        unsafe {
+            (*iosb).Anonymous.Status = STATUS_PENDING;
+        }
+
+        let ntdll = NtdllImports::get()?;
+        let result = unsafe {
+            ntdll.NtDeviceIoControlFile(
+                self.handle,
+                0,
+                ptr::null_mut(),
+                iosb.cast(),
+                iosb.cast(),
+                IOCTL_AFD_POLL,
+                poll_info.cast(),
+                size_of::<AfdPollInfo>() as u32,
+                poll_info.cast(),
+                size_of::<AfdPollInfo>() as u32,
+            )
+        };
+
+        match result {
+            STATUS_SUCCESS => Ok(()),
+            STATUS_PENDING => Err(io::ErrorKind::WouldBlock.into()),
+            status => {
+                let real_code = unsafe { ntdll.RtlNtStatusToDosError(status) };
+
+                Err(io::Error::from_raw_os_error(real_code as i32))
+            }
+        }
+    }
+
+    /// Cancel an ongoing poll operation.
+    ///
+    /// # Safety
+    ///
+    /// The poll operation must currently be in progress for this AFD.
+    pub(super) unsafe fn cancel(&self, packet: &T) -> io::Result<()> {
+        let ntdll = NtdllImports::get()?;
+
+        let result = {
+            // First, check if the packet is still in use.
+            let iosb = packet.as_ptr().cast::<IO_STATUS_BLOCK>();
+
+            if (*iosb).Anonymous.Status != STATUS_PENDING {
+                return Ok(());
+            }
+
+            // Cancel the packet.
+            let mut cancel_iosb = MaybeUninit::<IO_STATUS_BLOCK>::zeroed();
+
+            ntdll.NtCancelIoFileEx(self.handle, iosb, cancel_iosb.as_mut_ptr())
+        };
+
+        if result == STATUS_SUCCESS || result == STATUS_NOT_FOUND {
+            Ok(())
+        } else {
+            let real_code = ntdll.RtlNtStatusToDosError(result);
+
+            Err(io::Error::from_raw_os_error(real_code as i32))
+        }
+    }
+}
+
+/// A one-time initialization cell.
+struct OnceCell<T> {
```

Comment:
why do we roll our own `OnceCell` implementation here?

Reply:
We can't use `once_cell` since its MSRV is 1.56, and pulling in `async_lock::OnceCell` would add a few too many dependencies for my liking. I figured rolling my own is the best option.